source/s3_meta_request.c

```diff
@@ -120,75 +150,124 @@ static void s_s3_request_on_body(
     (void)meta_request;
     struct s3_meta_request_binding *request_binding = user_data;
 
+    bool report_progress;
+    if (s_record_progress(request_binding, (uint64_t)body->len, &report_progress)) {
+        return;
+    }
+    if (request_binding->recv_file) {
+        /* The callback will be invoked with the right order, so we don't need to seek first. */
+        if (fwrite((void *)body->ptr, body->len, 1, request_binding->recv_file) < body->len) {
+            /* fwrite failed */
+            /* TODO: return the error code back to native client. */
+            /* return aws_translate_and_raise_io_error(errno); */
+        }
+        if (!report_progress) {
+            return;
+        }
+    }
     /*************** GIL ACQUIRE ***************/
     PyGILState_STATE state;
+    PyObject *result = NULL;
     if (aws_py_gilstate_ensure(&state)) {
         return; /* Python has shut down. Nothing matters anymore, but don't crash */
     }
+    if (!request_binding->recv_file) {
+        result = PyObject_CallMethod(
+            request_binding->self_py,
+            "_on_body",
+            "(y#K)",
+            (const char *)(body->ptr),
+            (Py_ssize_t)body->len,
+            range_start);
 
-    PyObject *result = PyObject_CallMethod(
-        request_binding->self_py, "_on_body", "(y#K)", (const char *)(body->ptr), (Py_ssize_t)body->len, range_start);
-    if (!result) {
-        PyErr_WriteUnraisable(request_binding->self_py);
-        goto done;
+        if (!result) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+            goto done;
+        }
+        Py_DECREF(result);
+    }
+    if (report_progress) {
+        /* Hold the GIL before enterring here */
+        result =
+            PyObject_CallMethod(request_binding->self_py, "_on_progress", "(k)", request_binding->size_transferred);
+        if (!result) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+        } else {
+            Py_DECREF(result);
+        }
+        request_binding->size_transferred = 0;
     }
-    Py_DECREF(result);
-
 done:
     PyGILState_Release(state);
-    /*************** GIL RELEASE ***************/
+    return;
 }
 
-/* If the request has not finished, it will keep the request alive, until the finish callback invoked. So, we don't need
- * to clean anything from this call */
+/* If the request has not finished, it will keep the request alive, until the finish callback invoked. So, we don't
+ * need to clean anything from this call */
 static void s_s3_request_on_finish(
     struct aws_s3_meta_request *meta_request,
     const struct aws_s3_meta_request_result *meta_request_result,
     void *user_data) {
     (void)meta_request;
     struct s3_meta_request_binding *request_binding = user_data;
+    if (request_binding->input_body) {
+        aws_input_stream_destroy(request_binding->input_body);
```

Comment:
Noticed that this is the only place you're destroying the input_body.

Remember that a capsule can be in a partially-constructed state if aws_py_s3_client_make_meta_request() has an error.

You probably want to ALSO try and destroy it from s_destroy_if_ready() or the capsule destructor (you'd need to NULL out the variable here if you keep this destroy() in the on_finish)

I gotta go. But think about what will happen to ALL the stuff in your binding capsule if make_meta_request() fails at the last moment